### PR TITLE
Readme and JSDoc proofreading done

### PR DESCRIPTION
**Pull request by praveen5959**
Summary

READEME and JSDoc comments in the code have been proofread.

### CODE REVIEW

**Comment by balazsbotond** [comment]
Thank you for taking the time to proofread everything! Could you please also explain *why* you proposed each change?

**Comment by praveen5959** [comment]
In the index.ts file, some JSDoc comments didna have appropriate fields like return value so I have added those.
For some function, JSDoc comments were not present so I have added those as well.

In README file changes I have done to make the docs general and just first-person English.  And small other changes here and there.

File: .gitignore [hunk 1]
```diff
@@ -3,3 +3,4 @@
 /dist-test
 /.nyc_output
 /coverage
+.vscode
```
Comment (balazsbotond) [inline .gitignore:6]: Thanks for including this 👍

File: README.md [hunk 2]
```diff
@@ -52,7 +52,7 @@ Features:
 
 ## Why?
 
-When I need to call an HTTP API, I usually need to add dynamic parameters to the URL:
+When we call an HTTP API, we usually need to add dynamic parameters to the URL:
```
Comment (balazsbotond) [inline README.md]: I wrote this section in the first person because I want the README to be more personal. To be honest I would want to keep it that way; still, thanks for taking the time to read through it.

File: README.md [hunk 3]
```diff
@@ -63,12 +63,12 @@ function getUserPosts(id, blogId, limit, offset) {
 }
 ~~~
 
-As you can see, this minimal example is already rather hard to read. It is also incorrect:
+As we can see, this minimal example is already rather hard to read. It is also incorrect:
 
-- I forgot that there was a trailing slash at the end of the `API_URL` constant so the slash got duplicated (`https://api.example.com//users`)
+- One common issue is of trailing slashes, `API_URL` has a trailing slash in the example. This will lead to URL containing duplicated slashes (`https://api.example.com//users`).
```
Comment (balazsbotond) [inline README.md]: I agree that the phrase "the slash got duplicated" sounds unnatural. Can we just keep the second part ("This will lead to URL containing duplicated slashes") from here?

File: README.md [hunk 4]
```diff
@@ -63,12 +63,12 @@ function getUserPosts(id, blogId, limit, offset) {
...
 - The embedded values need to be escaped using `encodeURIComponent`
 
-I can use the built-in `URL` class to prevent duplicate slashes and `URLSearchParams` to escape the query string. But I still need to escape all path parameters manually.
+We can use the built-in `URL` class to prevent duplicate slashes and `URLSearchParams` to escape the query string. But we still need to escape all path parameters manually.
```
Comment (balazsbotond) [inline README.md]: As I said earlier, I would like to keep the first person voice. Could you please undo the changes related to it?

File: README.md [hunk 5]
```diff
@@ -103,29 +103,30 @@ The library handles:
...
 ```
 
 ### Usage with Node
 
-Node 10 and above are officially supported. Since the code uses the `URL` and `URLSearchParams` classes internally, which aren't available below v10, we cannot support those versions.
+Node 10 and above are officially supported. 
+Since the code uses the `URL` and `URLSearchParams` classes internally, which aren't available below v10, we cannot support those versions.
 
-If you want to build full URLs (most common use case):
+To build full URLs (most common use case):
```
Comment (balazsbotond) [inline README.md:117]: I agree with these 👍 Let's keep them

File: README.md [hunk 6]
```diff
@@ -103,29 +103,30 @@ The library handles:
...
 
-If you want to use any of the utility functions:
+To use any of the utility functions:
 
 ```ts
 const { query, subst, join } = require('urlcat');
 ```
 
-If you want to use everything:
+To use everything:
```
Comment (balazsbotond) [inline README.md]: Let's change this to "To use all exported functions:"

File: README.md [hunk 7]
```diff
@@ -262,7 +264,7 @@ Joins the two parts using exactly one separator. If a separator is present at th
 
 Thank you for using *urlcat*!
 
-If you need any help using this library, feel free to [create a GitHub issue](https://github.com/balazsbotond/urlcat/issues/new/choose) and ask your questions. I'll try to answer as quickly as possible.
+For any kind of help needed while using this library, feel free to [create a GitHub issue](https://github.com/balazsbotond/urlcat/issues/new/choose), and ask your questions. I'll try to answer as quickly as possible.
```
Comment (balazsbotond) [inline README.md]: I think we should keep the second person voice here. It's more personal and directly addresses the reader.

File: README.md [hunk 8]
```diff
@@ -272,7 +274,8 @@ Contributions of any kind (pull requests, bug reports, feature requests, documen
 
 Building the project should be quick and easy. If it isn't, it's the maintainer's fault. Please report any problems with building in a GitHub issue.
 
-You need to have a reasonably recent version of node.js to build *urlcat*. My node version is 12.18.3, npm is at 6.14.6.
+You need to have a reasonably recent version of node.js to build *urlcat*. 
+Tested on node version 12.18.3 and npm version 6.14.6.
```
Comment (balazsbotond) [inline README.md:278]: Thanks, this is much better this way.

File: src/index.ts [hunk 9]
```diff
@@ -6,7 +6,8 @@ export type ParamMap = Record<string, any>;
  * @param {String} baseTemplate a URL template that contains zero or more :params
  * @param {Object} params an object with properties that correspond to the :params
  *   in the base template. Unused properties become query params.
- * @returns a URL with path params substituted and query params appended
+ *
+ * @returns {String} a URL with path params substituted and query params appended
```
Comment (balazsbotond) [inline src/index.ts:10]: 👍

File: src/index.ts [hunk 10]
```diff
@@ -17,12 +18,13 @@ export type ParamMap = Record<string, any>;
 export default function urlcat(baseTemplate: string, params: ParamMap): string;
 
 /**
- * Concatenates the base URL and the path specified using '/' as a separator such
- * that if a '/' occurs at the concatenation boundary in either parameter, it is removed.
+ * Concatenates the base URL and the path specified using '/' as a separator.
+ * If a '/' occurs at the concatenation boundary in either parameter, it is removed.
+ *
+ * @param {String} baseUrl Base URL
+ * @param {String} path Path of URL or query :params
```
Comment (balazsbotond) [inline src/index.ts]: I wrote "first part" and "second part" here because I wanted to emphasize that you can break up the URL anywhere. The updated description of the `path` parameter isn't correct, query params aren't handled by this overload.

File: src/index.ts [hunk 11]
```diff
@@ -17,12 +18,13 @@ export type ParamMap = Record<string, any>;
...
  *
- * @param {String} baseUrl the first part of the URL
- * @param {String} path the second part of the URL
- * @returns {String} the result of the concatenation
+ * @returns {String} Concatenated URL
```
Comment (balazsbotond) [inline src/index.ts]: 👍

File: src/index.ts [hunk 12]
```diff
@@ -33,16 +35,17 @@ export default function urlcat(baseTemplate: string, params: ParamMap): string;
 export default function urlcat(baseUrl: string, path: string): string;
 
 /**
- * Concatenates the base URL and the path specified using '/' as a separator such
- * that if a '/' occurs at the concatenation boundary in either parameter, it is removed.
+ * Concatenates the base URL and the path specified using '/' as a separator.
+ * If a '/' occurs at the concatenation boundary in either parameter, it is removed.
```
Comment (balazsbotond) [inline src/index.ts:39]: Thanks, sounds much better this way!

File: src/index.ts [hunk 13]
```diff
@@ -33,16 +35,17 @@ export default function urlcat(baseTemplate: string, params: ParamMap): string;
...
  * Substitutes path parameters with the properties of the @see params object and appends
  * properties unused in the path as query params.
  *
- * @param {String} baseUrl the first part of the URL
- * @param {String} path the second part of the URL
- * @param {Object} params an object with properties that correspond to the :params
+ * @param {String} baseUrl Base URL
+ * @param {String} path Path of URL or query :params
```
Comment (balazsbotond) [inline src/index.ts]: Again, no query params allowed here

File: src/index.ts [hunk 14]
```diff
@@ -79,7 +82,8 @@ function urlcatImpl(pathTemplate: string, params: ParamMap, baseUrl?: string) {
  * Creates a query string from the object specified.
  *
  * @param {Object} params an object to convert into a query string.
- * @returns {String} a query string.
+ *
+ * @returns {String} Query string.
```
Comment (balazsbotond) [inline src/index.ts:86]: 👍

File: src/index.ts [hunk 15]
```diff
@@ -96,7 +100,8 @@ export function query(params: ParamMap): string {
  *
  * @param {String} template a string that contains :params.
  * @param {Object} params on object with keys that correspond to the params in the template.
- * @returns {String}
+ *
+ * @returns {String} Rendering path after substitution .
```
Comment (balazsbotond) [inline src/index.ts]: I would change this to "Rendered path after substitution."

File: src/index.ts [hunk 16]
```diff
@@ -109,6 +114,13 @@ export function subst(template: string, params: ParamMap): string {
   return renderedPath;
 }
 
+/**
+ * Creates the path and path :params
+ * @param {String} template a string that contains :params.
+ * @param {Object} params on object with keys that correspond to the params in the template.
+ *
+ * @returns {String, String} Rendering path and rendering :params.
+ */
```
Comment (balazsbotond) [inline src/index.ts]: This is an internal function that does not need a jsdoc string. Can you please remove it?

File: src/index.ts [hunk 17]
```diff
@@ -133,13 +145,14 @@ function path(template: string, params: ParamMap) {
 }
 
 /**
- * Joins two strings using a separator. If the separator occurs at the concatenation
- * boundary in either of the strings, it is removed. This prevents accidental duplication
- * of the separator.
+ * Joins two strings using a separator.
+ * If the separator occurs at the concatenation boundary in either parameter, it is removed.
```
Comment (balazsbotond) [inline src/index.ts]: "in either parameter" is incorrect; there are three parameters. I would keep my original version here.

File: src/index.ts [hunk 18]
```diff
@@ -133,13 +145,14 @@ function path(template: string, params: ParamMap) {
...
  *
- * @param {String} part1 the first string to join.
- * @param {String} separator the separator to use for joining.
- * @param {String} part2 the second string to join.
+ * @param {String} part1 First string.
+ * @param {String} separator Separator used for joining.
+ * @param {String} part2 Second string.
```
Comment (balazsbotond) [inline src/index.ts:147]: 👍

File: src/index.ts [hunk 19]
```diff
@@ -133,13 +145,14 @@ function path(template: string, params: ParamMap) {
...
+ *
+ * @returns {String} Joined string.
```
Comment (balazsbotond) [inline src/index.ts:149]: 👍

File: src/index.ts [hunk 20]
```diff
@@ -148,13 +161,21 @@ function path(template: string, params: ParamMap) {
...
-  return p1 === '' || p2 === ''
-    ? p1 + p2
-    : p1 + separator + p2;
+  const p1 = part1.endsWith(separator)
+    ? part1.slice(0, -separator.length)
+    : part1;
+  const p2 = part2.startsWith(separator)
+    ? part2.slice(separator.length)
+    : part2;
+  return p1 === "" || p2 === "" ? p1 + p2 : p1 + separator + p2;
```
Comment (balazsbotond) [inline src/index.ts]: I see why you reformatted the first two lines - they are slightly easier to read this way - but I see no reason to put the return statement on a single line. It's become very hard to read. Could you please format it again?

File: src/index.ts [hunk 21]
```diff
@@ -148,13 +161,21 @@ function path(template: string, params: ParamMap) {
...
 }
 
+/**
+ * Removes Null and undefined :params.
+ * @param params Object consisting of :params
+ *
+ * @returns {Object} Object of :params after removing Null and undefined values
+ */
```
Comment (balazsbotond) [inline src/index.ts]: This is an internal function that isn't exported so it doesn't need a jsdoc string.

File: src/index.ts [hunk 22]
```diff
@@ -164,6 +185,13 @@ function removeNullOrUndef(params: ParamMap) {
     }, {} as ParamMap);
 }
 
+/**
+ * Checks whether :param is Null  or undefined.
+ * @param v {any} :param
+ *
+ * @returns {Boolean} true -> :param is not Null or not undefined
+ * @returns {Boolean} false -> :param is Null or undefined
+ */
```
Comment (balazsbotond) [inline src/index.ts]: This is an internal function that isn't exported so it doesn't need a jsdoc string.

**Review by balazsbotond — CHANGES_REQUESTED** [review]
Thank you again for reading through everything and suggesting changes. I agree with many of your suggestions. I commented on each change and indicated whether I want to keep them. Would you please update the PR accordingly?

**Comment by praveen5959** [comment]
Sure
I will do it by today EOD.

**Comment by balazsbotond** [comment]
> Sure
> I will do it by today EOD.

Thank you!

**Comment by praveen5959** [comment]
I have updated the PR with the requested changes.

**Comment by balazsbotond** [comment]
Thank you! Merging 👍

**Comment by balazsbotond** [comment]
@all-contributors please add @praveen5959 for docs
